gemonlinedb/Makefile

```diff
@@ -28,6 +29,8 @@ Sources = \
     version.cc \
     XMLSerializationData.cc \
     XMLUtils.cc \
+    GEMOnlineDBManager.cc \
```

Comment:
Keep alphabetical order.

Reply:
Or use `wildcard`, unless there `.cc` files that **shouldn't** get built into the `.so` for some reason

Reply:
`wildcard` is bad because the build system will not attempt to rebuild in case a file is removed, and hence errors can remain unnoticed.

Reply:
Just want to understand the use case you think will fail, because I did some local testing and I ran what I think is your case is
```shell
make
rm src/file.c .
make
```

* With `wildcard` source determination, the second `make` doesn't trigger a rebuild (as probably expected)
* With explicit listing of sources (but not removing the removed source from the list), `make` fails with a `no rule to make target file.c, needed by file.o` (again, as expected, if a file is removed, but is still listed as "needed", `make` should fail because the developer made a mistake somewhere)
* Further removing `file.c` from the explicit list results in the same build sequence as with the `wildcard` method, i.e., no rebuild after removing the file

I fully agree that if code has been restructured in such a way that some source file is no longer present, a rebuild should be done, and I expect the developer to know they need to do a clean build in this case, but perhaps this wasn't exactly the use case you were driving at (possibly more related to the auto-generated code side?)

Reply:
```
make
git rm src/file.c
make # It works! Good
git commit
git push # CI broken
```

Reply:
Perhaps I'm failing to connect your concern with a specific situation in *this* codebase that you're considering, rather than in the abstract.
In your example above, does the following succeed or fail?
```
make
git rm src/file.c
make clean && make
```

I posit that this will **not** be a problem with the `wildcard` approach (but *could* still be a problem with the explicit listing approach if the `Makefile` is **not** updated when it should be) specifically since **every** CI build happens from a clean area, and the CI failing tells the developer about the mistake they made.

In any event, the CI isn't "broken", it's specifically doing exactly what it was intended to, no?

### Example project
#### `wildcard`
```shell
make
SOURCES   src/base.c src/a.c src/c.c src/b.c src/c/ext5.c src/b/ext3.c src/b/ext4.c src/a/ext1.c src/a/ext2.c
mkdir -p lib
gcc -std=c11 -fPIC -shared  -lm -Llib -Wl,-soname,base.so -o lib/base.so src/linux/base.o 
mkdir -p lib
gcc -std=c11 -fPIC -shared  -lm -Llib -Wl,-soname,a.so -o lib/a.so src/linux/a.o 
mkdir -p lib
gcc -std=c11 -fPIC -shared  -lm -Llib -Wl,-soname,c.so -o lib/c.so src/linux/c.o 
mkdir -p lib
gcc -std=c11 -fPIC -shared  -lm -Llib -Wl,-soname,b.so -o lib/b.so src/linux/b.o 
mkdir -p lib/c
gcc -std=c11 -fPIC -shared  -lm -Llib -Wl,-soname,ext5.so -o lib/c/ext5.so src/linux/c/ext5.o 
mkdir -p lib/b
gcc -std=c11 -fPIC -shared  -lm -Llib -Wl,-soname,ext3.so -o lib/b/ext3.so src/linux/b/ext3.o 
mkdir -p lib/b
gcc -std=c11 -fPIC -shared  -lm -Llib -Wl,-soname,ext4.so -o lib/b/ext4.so src/linux/b/ext4.o 
mkdir -p lib/a
gcc -std=c11 -fPIC -shared  -lm -Llib -Wl,-soname,ext1.so -o lib/a/ext1.so src/linux/a/ext1.o 
mkdir -p lib/a
gcc -std=c11 -fPIC -shared  -lm -Llib -Wl,-soname,ext2.so -o lib/a/ext2.so src/linux/a/ext2.o 

mv src/c/ext5.c .

make
SOURCES   src/base.c src/a.c src/c.c src/b.c src/b/ext3.c src/b/ext4.c src/a/ext1.c src/a/ext2.c
make: Nothing to be done for 'all'.

make clean
SOURCES   src/base.c src/a.c src/c.c src/b.c src/b/ext3.c src/b/ext4.c src/a/ext1.c src/a/ext2.c
rm -rf *~ src/linux/base.o src/linux/a.o src/linux/c.o src/linux/b.o src/linux/b/ext3.o src/linux/b/ext4.o src/linux/a/ext1.o src/linux/a/ext2.o src/linux/b/ext3.o src/linux/b/ext4.o src/linux/a/ext1.o src/linux/a/ext2.o lib/base.so lib/a.so lib/c.so lib/b.so lib/b/ext3.so lib/b/ext4.so lib/a/ext1.so lib/a/ext2.so src/linux/base.d src/linux/a.d src/linux/c.d src/linux/b.d src/linux/b/ext3.d src/linux/b/ext4.d src/linux/a/ext1.d src/linux/a/ext2.d src/linux/b/ext3.d src/linux/b/ext4.d src/linux/a/ext1.d src/linux/a/ext2.d

make
SOURCES   src/base.c src/a.c src/c.c src/b.c src/b/ext3.c src/b/ext4.c src/a/ext1.c src/a/ext2.c
mkdir -p lib
gcc -std=c11 -fPIC -shared  -lm -Llib -Wl,-soname,base.so -o lib/base.so src/linux/base.o 
mkdir -p lib
gcc -std=c11 -fPIC -shared  -lm -Llib -Wl,-soname,a.so -o lib/a.so src/linux/a.o 
mkdir -p lib
gcc -std=c11 -fPIC -shared  -lm -Llib -Wl,-soname,c.so -o lib/c.so src/linux/c.o 
mkdir -p lib
gcc -std=c11 -fPIC -shared  -lm -Llib -Wl,-soname,b.so -o lib/b.so src/linux/b.o 
mkdir -p lib/b
gcc -std=c11 -fPIC -shared  -lm -Llib -Wl,-soname,ext3.so -o lib/b/ext3.so src/linux/b/ext3.o 
mkdir -p lib/b
gcc -std=c11 -fPIC -shared  -lm -Llib -Wl,-soname,ext4.so -o lib/b/ext4.so src/linux/b/ext4.o 
mkdir -p lib/a
gcc -std=c11 -fPIC -shared  -lm -Llib -Wl,-soname,ext1.so -o lib/a/ext1.so src/linux/a/ext1.o 
mkdir -p lib/a
gcc -std=c11 -fPIC -shared  -lm -Llib -Wl,-soname,ext2.so -o lib/a/ext2.so src/linux/a/ext2.o 
```

#### explicit listing (not removed from `Makefile`)
```shell
make
SOURCES   src/a/ext1.c src/a/ext2.c src/a.c src/b/ext3.c src/b/ext4.c src/base.c src/b.c src/c/ext5.c src/c.c
mkdir -p lib/a
gcc -std=c11 -fPIC -shared  -lm -Llib -Wl,-soname,ext1.so -o lib/a/ext1.so src/linux/a/ext1.o 
mkdir -p lib/a
gcc -std=c11 -fPIC -shared  -lm -Llib -Wl,-soname,ext2.so -o lib/a/ext2.so src/linux/a/ext2.o 
mkdir -p lib
gcc -std=c11 -fPIC -shared  -lm -Llib -Wl,-soname,a.so -o lib/a.so src/linux/a.o 
mkdir -p lib/b
gcc -std=c11 -fPIC -shared  -lm -Llib -Wl,-soname,ext3.so -o lib/b/ext3.so src/linux/b/ext3.o 
mkdir -p lib/b
gcc -std=c11 -fPIC -shared  -lm -Llib -Wl,-soname,ext4.so -o lib/b/ext4.so src/linux/b/ext4.o 
mkdir -p lib
gcc -std=c11 -fPIC -shared  -lm -Llib -Wl,-soname,base.so -o lib/base.so src/linux/base.o 
mkdir -p lib
gcc -std=c11 -fPIC -shared  -lm -Llib -Wl,-soname,b.so -o lib/b.so src/linux/b.o 
mkdir -p lib/c
gcc -std=c11 -fPIC -shared  -lm -Llib -Wl,-soname,ext5.so -o lib/c/ext5.so src/linux/c/ext5.o 
mkdir -p lib
gcc -std=c11 -fPIC -shared  -lm -Llib -Wl,-soname,c.so -o lib/c.so src/linux/c.o 

mv src/c/ext5.c .

make
SOURCES   src/a/ext1.c src/a/ext2.c src/a.c src/b/ext3.c src/b/ext4.c src/base.c src/b.c src/c/ext5.c src/c.c
make: *** No rule to make target 'src/c/ext5.c', needed by 'src/linux/c/ext5.o'.  Stop.

make clean
SOURCES   src/a/ext1.c src/a/ext2.c src/a.c src/b/ext3.c src/b/ext4.c src/base.c src/b.c src/c/ext5.c src/c.c
rm -rf *~ src/linux/a/ext1.o src/linux/a/ext2.o src/linux/a.o src/linux/b/ext3.o src/linux/b/ext4.o src/linux/base.o src/linux/b.o src/linux/c/ext5.o src/linux/c.o src/linux/b/ext3.o src/linux/b/ext4.o src/linux/a/ext1.o src/linux/a/ext2.o lib/a/ext1.so lib/a/ext2.so lib/a.so lib/b/ext3.so lib/b/ext4.so lib/base.so lib/b.so lib/c/ext5.so lib/c.so src/linux/a/ext1.d src/linux/a/ext2.d src/linux/a.d src/linux/b/ext3.d src/linux/b/ext4.d src/linux/base.d src/linux/b.d src/linux/c/ext5.d src/linux/c.d src/linux/b/ext3.d src/linux/b/ext4.d src/linux/a/ext1.d src/linux/a/ext2.d

make
SOURCES   src/a/ext1.c src/a/ext2.c src/a.c src/b/ext3.c src/b/ext4.c src/base.c src/b.c src/c/ext5.c src/c.c
mkdir -p lib/a
gcc -std=c11 -fPIC -shared  -lm -Llib -Wl,-soname,ext1.so -o lib/a/ext1.so src/linux/a/ext1.o 
mkdir -p lib/a
gcc -std=c11 -fPIC -shared  -lm -Llib -Wl,-soname,ext2.so -o lib/a/ext2.so src/linux/a/ext2.o 
mkdir -p lib
gcc -std=c11 -fPIC -shared  -lm -Llib -Wl,-soname,a.so -o lib/a.so src/linux/a.o 
mkdir -p lib/b
gcc -std=c11 -fPIC -shared  -lm -Llib -Wl,-soname,ext3.so -o lib/b/ext3.so src/linux/b/ext3.o 
mkdir -p lib/b
gcc -std=c11 -fPIC -shared  -lm -Llib -Wl,-soname,ext4.so -o lib/b/ext4.so src/linux/b/ext4.o 
mkdir -p lib
gcc -std=c11 -fPIC -shared  -lm -Llib -Wl,-soname,base.so -o lib/base.so src/linux/base.o 
mkdir -p lib
gcc -std=c11 -fPIC -shared  -lm -Llib -Wl,-soname,b.so -o lib/b.so src/linux/b.o 
make: *** No rule to make target 'src/linux/c/ext5.o', needed by 'lib/c/ext5.so'.  Stop.
```

#### explicit listing (removed from `Makefile`)
```shell
make
SOURCES   src/a/ext1.c src/a/ext2.c src/a.c src/b/ext3.c src/b/ext4.c src/base.c src/b.c src/c/ext5.c src/c.c
mkdir -p lib/a
gcc -std=c11 -fPIC -shared  -lm -Llib -Wl,-soname,ext1.so -o lib/a/ext1.so src/linux/a/ext1.o 
mkdir -p lib/a
gcc -std=c11 -fPIC -shared  -lm -Llib -Wl,-soname,ext2.so -o lib/a/ext2.so src/linux/a/ext2.o 
mkdir -p lib
gcc -std=c11 -fPIC -shared  -lm -Llib -Wl,-soname,a.so -o lib/a.so src/linux/a.o 
mkdir -p lib/b
gcc -std=c11 -fPIC -shared  -lm -Llib -Wl,-soname,ext3.so -o lib/b/ext3.so src/linux/b/ext3.o 
mkdir -p lib/b
gcc -std=c11 -fPIC -shared  -lm -Llib -Wl,-soname,ext4.so -o lib/b/ext4.so src/linux/b/ext4.o 
mkdir -p lib
gcc -std=c11 -fPIC -shared  -lm -Llib -Wl,-soname,base.so -o lib/base.so src/linux/base.o 
mkdir -p lib
gcc -std=c11 -fPIC -shared  -lm -Llib -Wl,-soname,b.so -o lib/b.so src/linux/b.o 
mkdir -p lib/c
gcc -std=c11 -fPIC -shared  -lm -Llib -Wl,-soname,ext5.so -o lib/c/ext5.so src/linux/c/ext5.o 
mkdir -p lib
gcc -std=c11 -fPIC -shared  -lm -Llib -Wl,-soname,c.so -o lib/c.so src/linux/c.o 

mv src/c/ext5.c .
# modify Makefile

make
SOURCES   src/a/ext1.c src/a/ext2.c src/a.c src/b/ext3.c src/b/ext4.c src/base.c src/b.c src/c.c
make: Nothing to be done for 'all'.

make clean       
SOURCES   src/a/ext1.c src/a/ext2.c src/a.c src/b/ext3.c src/b/ext4.c src/base.c src/b.c src/c.c
rm -rf *~ src/linux/a/ext1.o src/linux/a/ext2.o src/linux/a.o src/linux/b/ext3.o src/linux/b/ext4.o src/linux/base.o src/linux/b.o src/linux/c.o src/linux/b/ext3.o src/linux/b/ext4.o src/linux/a/ext1.o src/linux/a/ext2.o lib/a/ext1.so lib/a/ext2.so lib/a.so lib/b/ext3.so lib/b/ext4.so lib/base.so lib/b.so lib/c.so src/linux/a/ext1.d src/linux/a/ext2.d src/linux/a.d src/linux/b/ext3.d src/linux/b/ext4.d src/linux/base.d src/linux/b.d src/linux/c.d src/linux/b/ext3.d src/linux/b/ext4.d src/linux/a/ext1.d src/linux/a/ext2.d

make
SOURCES   src/a/ext1.c src/a/ext2.c src/a.c src/b/ext3.c src/b/ext4.c src/base.c src/b.c src/c.c
mkdir -p lib/a
gcc -std=c11 -fPIC -shared  -lm -Llib -Wl,-soname,ext1.so -o lib/a/ext1.so src/linux/a/ext1.o 
mkdir -p lib/a
gcc -std=c11 -fPIC -shared  -lm -Llib -Wl,-soname,ext2.so -o lib/a/ext2.so src/linux/a/ext2.o 
mkdir -p lib
gcc -std=c11 -fPIC -shared  -lm -Llib -Wl,-soname,a.so -o lib/a.so src/linux/a.o 
mkdir -p lib/b
gcc -std=c11 -fPIC -shared  -lm -Llib -Wl,-soname,ext3.so -o lib/b/ext3.so src/linux/b/ext3.o 
mkdir -p lib/b
gcc -std=c11 -fPIC -shared  -lm -Llib -Wl,-soname,ext4.so -o lib/b/ext4.so src/linux/b/ext4.o 
mkdir -p lib
gcc -std=c11 -fPIC -shared  -lm -Llib -Wl,-soname,base.so -o lib/base.so src/linux/base.o 
mkdir -p lib
gcc -std=c11 -fPIC -shared  -lm -Llib -Wl,-soname,b.so -o lib/b.so src/linux/b.o 
mkdir -p lib
gcc -std=c11 -fPIC -shared  -lm -Llib -Wl,-soname,c.so -o lib/c.so src/linux/c.o 
```

Reply:
I think you're right and there's no difference in the case of a simple `Makefile`. One has to remember to run `make clean`, but that's a pitfall of using `make`.

I was confused because both Autotools, CMake and Meson advocate against `wildcard`. They can run something smarter than `make clean && make` automatically whenever their configuration is changed. Their detection logic fail if globbing is used.

Reply:
OK, got it :+1: